pygsti/tools/matrixtools.py

```diff
@@ -177,21 +215,22 @@ def nice_nullspace(m, tol=1e-7, orthogonalize=False):
     An matrix of shape (M,K) whose columns contain nullspace basis vectors.
     """
     nullsp = nullspace(m, tol)
-    nullsp_projector = _np.dot(nullsp, nullsp.conj().T)
-    keepers = []; current_rank = 0
-    for i in range(nullsp_projector.shape[1]):  # same as mx.shape[1]
-        rank = _np.linalg.matrix_rank(nullsp_projector[:, 0:i + 1], tol=tol)
-        if rank > current_rank:
-            keepers.append(i)
-            current_rank = rank
-    ret = _np.take(nullsp_projector, keepers, axis=1)
-
-    if orthogonalize:  # and not columns_are_orthogonal(ret):
-        ret, _ = _np.linalg.qr(ret)  # Gram-Schmidt orthogonalization
-
+    dim_ker = nullsp.shape[1]
+    if dim_ker == 0:
+        return nullsp  # empty 0-by-N array
+    _, _, p = _spl.qr(nullsp.T.conj(), mode='raw', pivoting=True)
+    ret = nullsp @ (nullsp.T[:, p[:dim_ker]]).conj()
+    # ^ That's equivalent to, but faster than:
+    #     nullsp_projector = nullsp @ nullsp.T.conj()
+    #     _, _, p = _spl.qr(nullsp_projector mode='raw', pivoting=True)
+    #     ret = nullsp_projector[:, p[:dim_ker]]
+
+    if orthogonalize:
+        ret, _ = _spl.qr(ret, mode='economic')
```

Comment:
Let's say ``m`` is $M \times N$ and $K =$``dim_ker`` is positive. Generically, we'll have $K = N - M$.

Computing the initial nullspace for the old implementation and the new implementation costs $O(M^2 N)$. In the new implementation that's the only operation which costs $O(M^2 N)$.

The original implementation had other expensive steps: SVDs of sizes at least $M \times \ell$ for $\ell \in \\{1,\ldots,\min(M,K)\\}$ and as large as $M \times \ell$ for $\ell \in \\{N-K+1,\ldots,N\\}$. The total complexity of these SVDs was around $O(M^4)$, if not larger.

Reply:
Is it the `pivoting=True` that essentially retains the same behavior as the old implementation of only taking the columns that increase the rank? I.e. pivoting will sort it/give us the indices such that taking the first `dim_ker` columns will return the matrix of the expected rank?

Reply:
Yup! Pivoting essentially sorts columns according to how they contribute to the numerical rank of the matrix. So if you know in advance that the matrix is rank $r$ (say, by having computed an SVD, or knowing that it's full-rank), then you can identify a set of columns that span its range by taking the first $r$ columns from pivoted QR decomposition.

The pivots chosen by pivoted QR are also invariant under left-multiplication by (partial) isometries, so I could pivot on ``nullsp.T.conj()`` instead of ``nullsp @ nullsp.T.conj()``.